Fix filters masking jinja2.exceptions.UndefinedError

s-hertel opened the pull request:
##### SUMMARY
Fixes #66977

I'm not sure of a good general way to test this to keep it from happening in the future.

##### ISSUE TYPE
- Bugfix Pull Request

s-hertel committed:
Fix filters

(cherry picked from commit cc7d2218854251a914e0c89c9c39116a4500f04f)

s-hertel committed:
Fix 3rd party plugins

s-hertel commented on lib/ansible/template/__init__.py:
```diff
@@ -252,7 +252,19 @@ def _unroll_iterator(func):
+        except (UndefinedError, AnsibleUndefinedVariable):
+            raise
+        except Exception as e:
+            if 'undefined' in to_native(e):
```
I'm not sure if this is a good idea, but I noticed that lookups don't have the same problem since we check that the terms are defined first and thought maybe we could retroactively check for filters. @nitzmahone What do you think?

mkrizek replied:
We have another wrapper that intercepts the result of a filter: `_wrap_native_text` so we would need to handle that there too, or create an additional wrapper just for this purpose. Also should we do the same for jinja tests?

nitzmahone replied:
That was the thing I was saying earlier- checking for/failing on undefined inputs upfront would probably be what most people would expect, but that'd be a breaking change. Even if we chose that breaking change, we'd probably need a way for filters that *want* to allow undefined inputs to do so (transparently to the caller of the filter).

And yeah, we probably need an analogue for tests, since the same problems exist.

nitzmahone replied:
I'm also unclear on why this needs to be here- the string check for `undefined` in the exception message feels dirty and a potential source of false-trip. What are we getting that's not in the `Undefined` class hierarchy- is it something we can fix elsewhere to get rid of this?

s-hertel replied:
@mkrizek Thanks for your insight!

@nitzmahone Yeah, I see what you were saying and that seems very featureish.

> I'm also unclear on why this needs to be here- the string check for undefined in the exception message feels dirty and a potential source of false-trip. What are we getting that's not in the Undefined class hierarchy- is it something we can fix elsewhere to get rid of this?

This (36db9950af4447b0d0c33ffc683c2d42c698d752) was the continuation of my thoughts on trying to fix 3rd party plugins. I kept the commits separate so it would be easy to drop if we don't want it - I just wanted to put it out there since `except Exception:` is common and was a suggested pattern in our plugin dev documentation.

nitzmahone replied:
Another option might be to build it the way we wish it were, then start a deprecation warning cycle on this bridging behavior (eg, do something like this and/or input inspection with an explicit optin/optout in plugin metadata, decorator, or whatever, then issue a dep warning on the exception catch or undefined inputs for plugins that haven't explicitly opted into receiving them).

:shrug:

s-hertel replied:
I had another idea for the 'undefined' hack, and I think it's in better shape now :crossed_fingers:

mkrizek commented on lib/ansible/template/__init__.py:
```diff
@@ -252,7 +252,19 @@ def _unroll_iterator(func):
+        except Exception as e:
+            if 'undefined' in to_native(e):
+                from jinja2.runtime import Undefined
+                undefined_input = any(isinstance(arg, (Undefined, AnsibleUndefined)) for arg in args)
```
Isn't checking for `Undefined` enough since the inheritance is `Undefined` -> `StrictUndefined` -> `AnsibleUndefined`?

mkrizek commented on lib/ansible/plugins/test/core.py:
```diff
@@ -170,6 +171,8 @@ def version_compare(value, version, operator='eq', strict=False):
         method = getattr(py_operator, operator)
         return method(Version(str(value)), Version(str(version)))
+    except UndefinedError as e:
+        raise errors.AnsibleUndefinedVariable('Version comparison: %s' % e)
```
At first I thought we could just re-raise the `UndefinedError` and let `do_template()` deal with that but converting to our exception early makes more sense. :+1:

s-hertel replied:
Yeah, I waffled back and forth about that too.

s-hertel replied:
I ended up changing this after all - with my latest change our filters are no longer intercepting the exception, just triggering the right one earlier

Joebortscheller commented:
What am I doing here

samccann commented on docs/docsite/rst/dev_guide/developing_plugins.rst:
```diff
@@ -36,6 +36,17 @@ You should return errors encountered during plugin execution by raising ``Ansibl
     except Exception as e:
         raise AnsibleError('Something happened, this was original exception: %s' % to_native(e))
 
+Since Ansible evaluates variables lazily, filter and test plugins should propagate the exceptions ``jinja2.exceptions.UndefinedError`` and ``AnsibleUndefinedVariable``.
```
is 'lazily' a.. erm... technical term? :-)  Or alternately, would most developers know what you mean by 'Ansible evaluates variables lazily...'

relrod replied:
@samccann It _is_ a technical term (meaning it only evaluates them when needed -- if it's never used, it's never evaluated).

I think most with programming experience would know what it means. But not sure if the average sysadmin would or not.

relrod replied:
"lazily" could just as well be replaced with "only when they are needed" or "only when they are used"

s-hertel committed:
Do the same for tests and filters

Update _wrap_native_text

Fix isinstance check

Add support to check for undefined variables up front

Replace 'undefined' hack by adhering to fail_on_undefined instead and add a deprecation warning

s-hertel committed:
elaborate on lazy evaluation

s-hertel committed:
Fix collections path - since builtins are excluded, undefined errors can be recovered more indiscriminately

samccann reviewed (approved):
docs portion LGTM

s-hertel committed:
Simplify decorator

Wrap jinja2 builtins to avoid 3rd party deprecation warnings
